replica-node/src/replica_util.rs

```diff
@@ -0,0 +1,268 @@
+                        },
+                        eligible_rpc_peers
+                            .iter()
+                            .map(|contact_info| contact_info.id)
+                            .collect::<Vec<_>>()
+                    );
+                }
+            }
+            eligible_rpc_peers
+        };
+
+        if !eligible_rpc_peers.is_empty() {
+            let contact_info =
+                &eligible_rpc_peers[thread_rng().gen_range(0, eligible_rpc_peers.len())];
+            return Some((contact_info.clone(), highest_snapshot_hash));
+        } else {
+            retry_reason = Some("No snapshots available".to_owned());
+        }
+    }
+}
+
+fn start_gossip_node(
+    identity_keypair: Arc<Keypair>,
+    cluster_entrypoints: &[ContactInfo],
+    ledger_path: &Path,
+    gossip_addr: &SocketAddr,
+    gossip_socket: UdpSocket,
+    expected_shred_version: Option<u16>,
+    gossip_validators: Option<HashSet<Pubkey>>,
+    should_check_duplicate_instance: bool,
+) -> (Arc<ClusterInfo>, Arc<AtomicBool>, GossipService) {
+    let contact_info = ClusterInfo::gossip_contact_info(
+        identity_keypair.pubkey(),
+        *gossip_addr,
+        expected_shred_version.unwrap_or(0),
+    );
+    let mut cluster_info = ClusterInfo::new(contact_info, identity_keypair);
+    cluster_info.set_entrypoints(cluster_entrypoints.to_vec());
+    cluster_info.restore_contact_info(ledger_path, 0);
+    let cluster_info = Arc::new(cluster_info);
+
+    let gossip_exit_flag = Arc::new(AtomicBool::new(false));
+    let gossip_service = GossipService::new(
+        &cluster_info,
+        None,
+        gossip_socket,
+        gossip_validators,
+        should_check_duplicate_instance,
+        &gossip_exit_flag,
+    );
+    info!("Started gossip node");
+    info!(
+        "The cluster contact info:\n{}",
+        cluster_info.contact_info_trace()
+    );
+
+    (cluster_info, gossip_exit_flag, gossip_service)
+}
+
+// Get the RPC peer info given the peer's Pubkey
+// Returns the ClusterInfo, the peer's ContactInfo and the initial snapshot info
+pub fn get_rpc_peer_info(
+    identity_keypair: Keypair,
+    cluster_entrypoints: &[ContactInfo],
+    ledger_path: &Path,
+    node: &Node,
+    expected_shred_version: Option<u16>,
+    peer_pubkey: &Pubkey,
+    snapshot_output_dir: &Path,
+) -> (Arc<ClusterInfo>, ContactInfo, Option<(Slot, Hash)>) {
+    let identity_keypair = Arc::new(identity_keypair);
+
+    let gossip = Some(start_gossip_node(
+        identity_keypair,
+        cluster_entrypoints,
+        ledger_path,
+        &node.info.gossip,
+        node.sockets.gossip.try_clone().unwrap(),
+        expected_shred_version,
+        None,
+        true,
+    ));
+
+    let rpc_node_details = get_rpc_peer_node(
+        &gossip.as_ref().unwrap().0,
+        cluster_entrypoints,
+        expected_shred_version,
+        peer_pubkey,
+        snapshot_output_dir,
+    );
+    let rpc_node_details = rpc_node_details.unwrap();
```

Comment:
Is this unwrap safe?

Reply:
The way it is written -- yes -- it would have quit otherwise